### ISSUE TITLE
Multiprocess encoding for speed

### ISSUE THREAD

**Issue by davidmezzetti** [MinishLab/model2vec#139]
While Model2Vec is fast, I'm wondering if it can be even faster 🚀

The Sentence-Transformers library has a [multiprocess encoding method](https://github.com/UKPLab/sentence-transformers/blob/master/examples/applications/computing-embeddings/computing_embeddings_multi_gpu.py) that is useful for multiple GPUs. 

I was planning to add something similar to txtai for Model2Vec but first I wanted to check if perhaps it made sense to add it directly to the library. 

So I tested encoding **10 million inputs** with the standard `encode` vs a multiprocessing `encode`. The standard encode method took **2 minutes**, a multiprocessing encode method on my machine took **20 seconds**. 

## Standard
```python
from model2vec import StaticModel

model = StaticModel.from_pretrained("minishlab/potion-base-8M")
print(model.encode(["1000"] * 10000000, batch_size=1024).shape)
```

```bash
$ time python standard.py 
(10000000, 256)

real	2m13.506s
user	3m29.900s
sys	0m11.513s
```

## Multiprocess
```python
import queue
import os

from multiprocessing import Process, Queue

import numpy as np

from model2vec import StaticModel


class Execute:
    @staticmethod
    def worker(model, inputs, outputs):
        while True:
            try:
                outputs.put(model.encode(inputs.get()))
            except queue.Empty:
                break

    @staticmethod
    def start(model, size=None):
        # Create queues
        inputs, outputs = Queue(), Queue()

        # Start worker processes
        processes = []
        for _ in range(size if size else os.cpu_count()):
            process = Process(target=Execute.worker, args=(model, inputs, outputs))
            process.start()
            processes.append(process)

        return processes, inputs, outputs

    @staticmethod
    def encode(data, pool, batch):
        _, inputs, outputs = pool

        # Chunk and add data to be processed
        chunks = 0
        for x in range(0, len(data), batch):
            inputs.put(data[x : x + batch])
            chunks += 1

        # Collect results
        embeddings = []
        for _ in range(chunks):
            embeddings.append(outputs.get())

        # Combine into single array and return
        return np.concatenate(embeddings)

    @staticmethod
    def stop(pool):
        if pool:
            processes, inputs, outputs = pool

            for process in processes:
                process.terminate()

            for process in processes:
                process.join()
                process.close()

            inputs.close()
            outputs.close()

if __name__ == "__main__":
    pool = None
    try:
        model = StaticModel.from_pretrained("minishlab/potion-base-8M")
        pool = Execute.start(model)
        print(Execute.encode(["1000"] * 10000000, pool, 1024).shape)

    finally:
        Execute.stop(pool)
```

```bash
$ time python multiprocess.py 
(10000000, 256)

real	0m20.879s
user	3m35.571s
sys	0m29.292s

```

Any interest?

**Comment by stephantul** [comment]
Hello @davidmezzetti ,

Thanks for the cool idea and nice implementation! We already checked this out a while ago, and decided against it, because the speed increase is minimal on real data, and could lead to unexpected things, more code to maintain, etc. etc.

Here's an example using Wikipedia:

```python
from datasets import load_dataset
from model2vec import StaticModel

ds = load_dataset("wikimedia/wikipedia",  
                              data_files="20231101.en/train-00000-of-00041.parquet")["train"]

model = StaticModel.from_pretrained("minishlab/potion-base-8M")
print(model.encode(ds["text"], batch_size=1024).shape)
```

This leads to the following timings on my macbook pro:
```
149.27s user 
3.67s system 
599% cpu 
25.534 total
```

Using the same dataset in your multiprocessing script leads to the following timings:
```
140.16s user 
7.71s system 
704% cpu 
20.999 total
```

As you can see, this is probably just because it uses 7 cores instead of 6. The reason for the discrepancy between this timing and yours is that about 80-95% of our `encode` time is spent in tokenization. But huggingface's tokenizers library already multiprocesses tokenization under the hood. 

As your example benchmark only uses strings comprising of single tokens, the real bottleneck kind of got hidden.

Now, I'm not sure whether a ~20% speedup is something we should turn down, but I would like to get a better view of the trade-offs, especially on other platforms (I've only tested this on my own MacBook). Let me know what you think, and thanks again for the idea!

Stéphan

**Comment by davidmezzetti** [comment]
Thank you for the response @stephantul. Glad to hear you guys were thinking of this!

I'm a Linux user. Linux uses `fork` for multiprocessing and macOS uses `spawn` by default.  [See this link for more on that](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods).

For example, when I run the scripts above for that dataset, I get this (I downloaded the files first to not count that time).

**Standard**
```bash
$ time python standard.py 
(156289, 256)

real	0m21.185s
user	2m18.819s
sys	0m0.944s
```

**Multiprocess**
```bash
$ time python mp.py 
(156289, 256)

real	0m10.992s
user	2m28.220s
sys	0m3.603s
```

`spawn` has to make a full copy of the process whereas `fork` doesn't. So I think that is part of the difference. If the scripts are modified to measure the time elapsed just for the encode call.

**Standard**
```python
import time

from model2vec import StaticModel

from datasets import load_dataset

ds = load_dataset("wikimedia/wikipedia", data_files="20231101.en/train-00000-of-00041.parquet")["train"]

model = StaticModel.from_pretrained("minishlab/potion-base-8M")

start = time.time()
print(model.encode(ds["text"],  batch_size=1024).shape)
print("Elapsed", time.time() - start)
```

```bash
$ time python standard.py
(156289, 256)
Elapsed 18.192959547042847

real	0m21.125s
user	2m19.078s
sys	0m0.990s
```

Now if I force my machine to spawn new processes vs fork.

**Multiprocess**
```python
import queue
import os
import time

import multiprocessing
from multiprocessing import Process, Queue

import numpy as np

from datasets import load_dataset
from model2vec import StaticModel


class Execute:
    @staticmethod
    def worker(model, inputs, outputs):
        while True:
            try:
                outputs.put(model.encode(inputs.get()))
            except queue.Empty:
                break

    @staticmethod
    def start(model, size=None):
        # Create queues
        inputs, outputs = Queue(), Queue()

        # Start worker processes
        processes = []
        for _ in range(size if size else os.cpu_count()):
            process = Process(target=Execute.worker, args=(model, inputs, outputs))
            process.start()
            processes.append(process)

        return processes, inputs, outputs

    @staticmethod
    def encode(data, pool, batch):
        _, inputs, outputs = pool

        # Chunk and add data to be processed
        chunks = 0
        for x in range(0, len(data), batch):
            inputs.put(data[x : x + batch])
            chunks += 1

        # Collect results
        embeddings = []
        for _ in range(chunks):
            embeddings.append(outputs.get())

        # Combine into single array and return
        return np.concatenate(embeddings)

    @staticmethod
    def stop(pool):
        if pool:
            processes, inputs, outputs = pool

            for process in processes:
                process.terminate()

            for process in processes:
                process.join()
                process.close()

            inputs.close()
            outputs.close()

if __name__ == "__main__":
    multiprocessing.set_start_method("spawn")

    pool = None
    try:
        model = StaticModel.from_pretrained("minishlab/potion-base-8M")
        pool = Execute.start(model)

        ds = load_dataset("wikimedia/wikipedia", data_files="20231101.en/train-00000-of-00041.parquet")["train"]

        start = time.time()
        print(Execute.encode(ds["text"], pool, 1024).shape)
        print("Elapsed", time.time() - start)

    finally:
        Execute.stop(pool)
```

```bash
$ time python mp.py 
(156289, 256)
Elapsed 8.244934797286987

real	0m21.132s
user	3m26.072s
sys	0m5.452s
```

So while the total process time is the same in this case (more time spent on spawning the processes), the encoding time is still significantly less (8s vs 18s). I believe macOS does support `fork`, so you could see for yourself on that. Nonetheless, the encoding time should still be significantly less. For long running processes that are encoding a lot of data, the time savings could be significant once the initial price is paid to spawn the processes.

For example, if the scripts above are expanded to process the first 4 wikipedia files (mp.py still uses spawn).

**Standard**
```bash
$ time python standard.py 
(625156, 256)
Elapsed 69.826988697052

real	1m13.018s
user	8m30.853s
sys	0m4.012s

```

**Multiprocess**
```bash
$ time python mp.py 
(625156, 256)
Elapsed 33.545076847076416

real	0m46.655s
user	10m43.425s
sys	0m11.786s
```

At least in my tests, the speedup is more on the order of 100% vs 20%. It would be interesting to see if you could reproduce this with your setup.

**Comment by bhavnicksm** [comment]
Hey @davidmezzetti!

This is great; thank you! I also tested on Colab and on my Mac and was wondering why I am getting speed-ups on Colab and not on my Mac. Thanks for the explanation.

I ran the "War and Peace" book 10,000 times with and without multiprocessing. 
<img width="585" alt="image" src="https://github.com/user-attachments/assets/7e931419-510a-4f28-b4fe-e28b53e094ea" />

<img width="497" alt="image" src="https://github.com/user-attachments/assets/1dc8f9f0-e99c-42f7-a4ce-382eac608825" />

And also, with MPIRE

<img width="642" alt="image" src="https://github.com/user-attachments/assets/26d0cbfc-9af1-4f26-b82f-2d7ca4026557" />


And, we can see there being about a 16.2s vs. 3.5s, which is an order of 4.65x order difference~!

(Just wanted to highlight this; I hope it is relevant~)

Thanks! ☺️

**Comment by Pringled** [comment]
Heya @davidmezzetti, we've done a bit more testing and we indeed saw some major speedups, really cool, thanks for the great suggestion! We've just merged a PR that adds multiprocessing to model2vec: https://github.com/MinishLab/model2vec/pull/141. We will be sure to mention you when we release it (and if you're interested you can already try it on main)!

**Comment by davidmezzetti** [comment]
Very cool, glad to hear it!

One question, does ProcessParallel live beyond calls? I know there is a price to pay for spinning up the processes, especially on MacOS. In txtai, I batch encode for various reasons, so there normally wouldn't be a call with a batch greather than 1024-2048. In the example code above, the processes were created once and terminated once (like what SentenceTransformers does).

If ProcessParallel does maintain a pool of processes (which from what I've read it does), would it be possible to add a parameter for that multiprocessing threshold of 10000?

**Comment by Pringled** [comment]
That's a good point! We've just updated this in https://github.com/MinishLab/model2vec/pull/142 (adding a multiprocessing_threshold parameter).

**Comment by davidmezzetti** [comment]
I appreciate all the work to integrate this in! I'll try it out and report back.

**Comment by chapter544** [comment]
sorry to hijack the discussion. I am integrating model2vec + gunicorn to serve as an API via fastapi and I meet with an issue about speed.

When I created a test script: 1) load the model, 2) encode a list of 1000 lines of text (200 word per line), the timing is very good: ~100-200ms. However, when I wrap it in a gunicorn process (to serve as an API), it takes 10-20x slower to encode the same amount of text. Do you have any pointers?

Thanks,

**Comment by stephantul** [comment]
Hello @chapter544,

That's unfortunate! 

Are you using the version that already includes multiprocessing? If so, could you turn it off? If that isn't the issue, I recommend you set:

```bash
TOKENIZERS_PARALLELISM=false
```

In your environment before starting gunicorn. What I think is happening is that the parallelism in tokenizers is interfering with the workers in gunicorn. Could you verify that this is the case?

If my pointers above don't solve it, please make a new issue, so we can tackle it separately. Thanks!

**Comment by davidmezzetti** [comment]
Thanks again for adding this change in. Please feel free to close this issue.

**Comment by do-me** [comment]
Just for reference #56 where we talk about performance & multiprocessing.
